File: LC0337_House_Robber_3/house_robber_3.cpp
#include <unordered_map>
#include <algorithm>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    // tracks max amount of money we can rob in the subtree rooted at each node, if we do rob the node
    unordered_map<TreeNode*, int> max_rob; 

    // tracks max amount of money we can rob in the subtree rooted at each node, if we do NOT rob the node
    unordered_map<TreeNode*, int> max_no_rob;


    bool isLeaf(const TreeNode* const node) const
    {
        return node != nullptr && node->left == nullptr && node->right == nullptr;
    }

    void rRob(TreeNode* root)
    {
        if (root == nullptr)
        {
            return;
        }
        if (this->isLeaf(root)) // base case
        {
            this->max_rob[root] = root->val;
            this->max_no_rob[root] = 0;
        }
        else // recursion
        {
            this->rRob(root->left);
            this->rRob(root->right);
            int a = root->left == nullptr ? 0 : this->max_rob[root->left];
            int b = root->right == nullptr ? 0 : this->max_rob[root->right];
            int c = root->left == nullptr ? 0 : this->max_no_rob[root->left];
            int d = root->right == nullptr ? 0 : this->max_no_rob[root->right];
            
            // If we rob root, then we cannot rob either of its children.
            this->max_rob[root] = root->val + c + d;

            // If we don't rob root, then we consider robbings its children.
            this->max_no_rob[root] = max(a, c) + max(b, d);
        }
    }

    int rob(TreeNode* root) {
        this->rRob(root);
        return max(this->max_rob[root], this->max_no_rob[root]);
    }
};

File: LC0337_House_Robber_3/house_robber_3.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def __init__(self):
        # maps node to max amount of money we can make from robbing the subtree rooted at the node, if we do rob this node
        self.max_rob = {} 
        # maps node to max amount of money we can make from robbing the subtree rooted at the node, if we do NOT rob this node
        self.max_no_rob = {} 

    def is_leaf(self, node: TreeNode) -> bool:
        return node and (not node.left) and (not node.right)

    def rRob(self, root: TreeNode):
        if not root:
            return
        if self.is_leaf(root):
            self.max_rob[root] = root.val
            self.max_no_rob[root] = 0
        else:
            self.rRob(root.left)
            self.rRob(root.right)
            a = self.max_no_rob[root.left] if root.left else 0
            b = self.max_no_rob[root.right] if root.right else 0
            c = self.max_rob[root.left] if root.left else 0
            d = self.max_rob[root.right] if root.right else 0

            # If we rob root, then we cannot rob either of its children.
            self.max_rob[root] = root.val + a + b

            # If we don't rob root, then we have the choice of either robbing or not robbing either of its children.
            self.max_no_rob[root] = max(a, c) + max(b, d)

    def rob(self, root: Optional[TreeNode]) -> int:
        self.rRob(root)
        return max(self.max_rob[root], self.max_no_rob[root])